python/tchannel/tornado/hyperbahn.py

```diff
@@ -58,34 +138,16 @@ def advertise(tchannel, service, routers):
         # TChannel already knows about some of the routers.
         tchannel.peers.get(router)
 
-    @tornado.gen.coroutine
-    def _register():
-        response = yield tchannel.request(service='hyperbahn').send(
-            arg1='ad',  # advertise
-            arg2='',
-            arg3=json.dumps({
-                'services': [
-                    {
-                        'serviceName': service,
-                        'cost': 0,
-                    }
-                ]
-            }),
-            headers={'as': 'json'},
-        )
+    ad_future = _first_advertise(tchannel, service)
 
-        if response.code not in ErrorCode:
-            # re-register every ``DEFAULT_TTL`` seconds
-            tornado.ioloop.IOLoop.current().call_later(
-                delay=DEFAULT_TTL,
-                callback=_register,
-            )
-            raise tornado.gen.Return(response)
+    def first_ad_callback(f):
+        if f.exception():
+            log.error('Failed to register with Hyperbahn forever.')
         else:
-            log.error('Failed to register with Hyperbahn: %s', response)
-            raise NotImplementedError  # TODO figure out behavior here
-
-    return _register()
+            _regular_advertise(tchannel, service)
 
+    with timeout(ad_future, DEFAULT_FIRST_ADVERTISE_TIME):
```

Comment:
We discussed this before. It's because with_timeout throws tornado timeout exception. I use Tchannel timeout exception. If we use tornado's , we need another exception translate.